src/main/java/com/google/devtools/build/lib/analysis/constraints/IncompatibleDeterminingHelper.java

```diff
@@ -54,141 +53,161 @@
 import com.google.devtools.build.lib.packages.RuleVisibility;
 import com.google.devtools.build.lib.packages.Target;
 import com.google.devtools.build.lib.skyframe.ConfiguredTargetAndData;
+import com.google.devtools.build.lib.skyframe.ConfiguredTargetKey;
 import com.google.devtools.build.lib.skyframe.RuleConfiguredTargetValue;
-import com.google.devtools.build.lib.skyframe.UnloadedToolchainContext;
 import com.google.devtools.build.lib.util.OrderedSetMultimap;
 import com.google.devtools.build.skyframe.SkyFunction.Environment;
+import com.google.devtools.build.skyframe.SkyframeLookupResult;
 import java.util.List;
 import javax.annotation.Nullable;
 
-public class RuleConstraintSemantics {
-  @AutoValue
-  public abstract static class IncompatibleTargetCreationResult {
-    private static IncompatibleTargetCreationResult create(
-        boolean needsMoreConfiguredTargets, RuleConfiguredTargetValue incompatibleTarget) {
-      return new AutoValue_RuleConstraintSemantics_IncompatibleTargetCreationResult(
-          needsMoreConfiguredTargets, incompatibleTarget);
-    }
-
-    public abstract boolean needsMoreConfiguredTargets();
-
-    @Nullable
-    public abstract RuleConfiguredTargetValue incompatibleTarget();
-  }
-
+/**
+ * Helpers for creating configured targets that are incompatible.
+ *
+ * <p>A target is considered incompatible if any of the following applies:
+ *
+ * <ol>
+ *   <li>The target's <code>target_compatible_with</code> attribute specifies a constraint that is
+ *       not present in the target platform.
+ *   <li>One or more of the target's dependencies is incompatible.
+ * </ol>
+ *
+ * The intent of these helpers is that they get called as early in the analysis phase as possible.
```

Comment:
I assume that this is why there are two methods, so that the direct check can be done ASAP. Call that out explicitly, I was initially looking to see if it was possible to unify the methods.

Reply:
Done